workload/remastering_workload.cpp

```diff
@@ -0,0 +1,196 @@
+
+#include <algorithm>
+#include <iomanip>
+#include <fcntl.h>
+#include <random>
+#include <sstream>
+#include <unordered_set>
+#include <vector>
+
+#include <glog/logging.h>
+
+#include "common/offline_data_reader.h"
+#include "common/proto_utils.h"
+#include "proto/offline_data.pb.h"
+
+using std::discrete_distribution;
+using std::shuffle;
+using std::uniform_int_distribution;
+using std::string;
+using std::vector;
+using std::unordered_set;
+
+namespace slog {
+namespace {
+
+constexpr char MH_PCT[] = "mh";
+constexpr char MH_NUM_HOMES[] = "mh_homes";
+constexpr char MP_PCT[] = "mp";
+constexpr char MP_NUM_PARTS[] = "mp_parts";
+constexpr char NUM_RECORDS[] = "num_records";
+constexpr char NUM_WRITES[] = "num_writes";
+constexpr char VALUE_SIZE[] = "value_size";
+constexpr char REMASTER_GAP[] = "remaster_gap";
+
+const RawParamMap DEFAULT_PARAMS = {
+  { MH_PCT, "0" },
+  { MH_NUM_HOMES, "2" },
+  { MP_PCT, "0" },
+  { MP_NUM_PARTS, "2" },
+  { NUM_RECORDS, "10" },
+  { NUM_WRITES, "2" },
+  { VALUE_SIZE, "100" }, // bytes
+  { REMASTER_GAP, "50" }
+};
+
+} // namespace
+
+RemasteringWorkload::RemasteringWorkload(
+    ConfigurationPtr config,
+    const string& data_dir,
+    const string& params_str)
+  : WorkloadGenerator(DEFAULT_PARAMS, params_str),
+    config_(config),
+    partition_to_key_lists_(config->GetNumPartitions()),
+    client_txn_id_counter_(0) {
+
+  for (auto& key_lists : partition_to_key_lists_) {
+    for (uint32_t rep = 0; rep < config->GetNumReplicas(); rep++) {
+      // TODO (ctring): initialize each key list with some hot keys
+      key_lists.emplace_back();
+    }
+  }
+
+  // Load and index the initial data
+  for (uint32_t partition = 0; partition < config->GetNumPartitions(); partition++) {
+    auto data_file = data_dir + "/" + std::to_string(partition) + ".dat";
+    auto fd = open(data_file.c_str(), O_RDONLY);
+    if (fd < 0) {
+      LOG(FATAL) << "Error while loading \"" << data_file << "\": " << strerror(errno);
+    }
+
+    OfflineDataReader reader(fd);
+    LOG(INFO) << "Loading " << reader.GetNumDatums() << " datums from " << data_file;
+    while (reader.HasNextDatum()) {
+      auto datum = reader.GetNextDatum();
+      CHECK_LT(datum.master(), config->GetNumReplicas())
+          << "Master number exceeds number of replicas";
+
+      partition_to_key_lists_[partition][datum.master()].AddKey(datum.key());
+    }
+    close(fd);
+  }
+}
+
+std::pair<Transaction*, TransactionProfile>
+RemasteringWorkload::NextTransaction() {
+  TransactionProfile pro;
+
+  pro.client_txn_id = client_txn_id_counter_;
+
+  Transaction* txn;
+  if (client_txn_id_counter_ % params_.GetUInt32(REMASTER_GAP) == 0) {
+    txn = MakeRemasterTransaction(pro);
+  } else {
+    txn = MakeNormalTransaction(pro);
+  }
+
+  txn->mutable_internal()->set_id(client_txn_id_counter_);
+
+  client_txn_id_counter_++;
+
+  return {txn, pro};
+}
+
+Transaction* RemasteringWorkload::MakeNormalTransaction(TransactionProfile& pro) {
```

Comment:
@ctring To avoid this duplication is it a good idea to extend the basic workload?

Reply:
Sounds good to me. You might also need to change `DEFAULT_PARAMS` from a const map to value returned from a virtual function so that you can override it.